src/xds/provider/agones.rs

```diff
@@ -0,0 +1,398 @@
+                        }),
+                        ..<_>::default()
+                    })),
+                    metadata: Some(Metadata {
+                        filter_metadata: HashMap::from([(
+                            String::from(QUILKIN_URL),
+                            Struct {
+                                fields: BTreeMap::from([(String::from(QUILKIN_TOKEN_KEY), tokens)]),
+                            },
+                        )]),
+                        ..<_>::default()
+                    }),
+                    ..<_>::default()
+                }
+            })
+            .collect();
+
+        Ok(Cluster {
+            name: DEFAULT_CLUSTER_NAME.into(),
+            load_assignment: Some(ClusterLoadAssignment {
+                cluster_name: DEFAULT_CLUSTER_NAME.into(),
+                endpoints: vec![LocalityLbEndpoints {
+                    lb_endpoints,
+                    ..<_>::default()
+                }],
+                ..<_>::default()
+            }),
+            cluster_discovery_type: Some(ClusterDiscoveryType::Type(DiscoveryType::Static as i32)),
+            ..<_>::default()
+        })
+    }
+}
+
+#[tonic::async_trait]
+impl crate::xds::DiscoveryServiceProvider for AgonesProvider {
+    async fn discovery_request(
+        &self,
+        _node_id: &str,
+        version: u64,
+        kind: ResourceType,
+        _names: &[String],
+    ) -> Result<DiscoveryResponse, tonic::Status> {
+        let resources = match kind {
+            ResourceType::Endpoint => {
+                let cluster = self.get_cluster().await?;
+                let mut buf = Vec::new();
+                buf.reserve(cluster.encoded_len());
+
+                cluster.encode(&mut buf).unwrap();
+
+                vec![Any {
+                    type_url: crate::xds::ENDPOINT_TYPE.into(),
+                    value: buf,
+                }]
+            }
+            ResourceType::Listener => {
+                let listener = self.get_listener().await?;
+                let mut buf = Vec::new();
+                buf.reserve(listener.encoded_len());
+
+                listener.encode(&mut buf).unwrap();
+
+                vec![Any {
+                    type_url: kind.type_url().into(),
+                    value: buf,
+                }]
+            }
+            kind => {
+                return Err(tonic::Status::internal(format!(
+                    "Quilkin currently does not support {} requests",
+                    kind.type_url()
+                )))
+            }
+        };
+
+        Ok(DiscoveryResponse {
+            version_info: version.to_string(),
+            resources,
+            type_url: kind.type_url().into(),
+            ..<_>::default()
+        })
+    }
+}
+
+#[derive(kube::CustomResource, Clone, Debug, Deserialize, Serialize, JsonSchema)]
+#[kube(
+    group = "agones.dev",
+    version = "v1",
+    kind = "GameServer",
+    status = "GameServerStatus",
+    namespaced
+)]
```

Comment:
A lot of these types and information is also implemented in [`rymder`](https://github.com/EmbarkStudios/rymder/blob/50ee2876d71f4dd4a82245e54023edc5b46fa5be/src/sdk/gameserver.rs). Is it a bad idea to consider patching `rymder` to make it more mature if needed and then depend on that crate here instead of duplicating everything? Or if not maybe the [agones Rust SDK](https://github.com/googleforgames/agones/tree/main/sdks/rust)?

Reply:
I think that would be up to @markmandel. My understanding from going through the Rust SDK and the other language SDKs. Is that some of those types are available, but that is for interacting with Agones through its GRPC API, rather than through k8s which is what we're doing.

I've no problem pushing those types elsewhere, but we'd want to ensure that they're compatible with `kube` which is also why I didn't use `rymder`/Rust SDK.

Reply:
+1 to what @XAMPPRocky said - the `rymder` and Agones SDK definitely has similar structures (because they are based on the same underlying data structures), but are aimed at different use cases. I think (at least for now), it's better to keep them split, so they don't have dependencies on each other - also, interacting with a K8s `GameServer` resource doesn't need the gRPC SDK integration for Agones, so it feels to me like something we should keep seperate.